pkg/river/diag/diag.go

```diff
@@ -0,0 +1,84 @@
+// Package diag exposes error types used throughout River and a method to
+// pretty-print them to the screen.
+package diag
+
+import (
+	"fmt"
+
+	"github.com/grafana/agent/pkg/river/token"
+)
+
+// Severity denotes the severity level of a diagnostic. The zero value of
+// severity is invalid.
+type Severity int
+
+// Supported severity levels.
+const (
+	SeverityLevelWarn Severity = iota + 1
+	SeverityLevelError
+)
+
+// Diagnostic is an individual diagnostic message. Diagnostic messages can have
+// different levels of severities.
+type Diagnostic struct {
+	// Severity holds the severity level of this Diagnostic.
+	Severity Severity
+
+	// StartPos refers to a position in a file where this Diagnostic starts.
+	StartPos token.Position
+
+	// EndPos refers to an optional position in a file where this Diagnostic
+	// ends. If EndPos is the zero value, the Diagnostic should be treated as
+	// only covering a single character (i.e., StartPos == EndPos).
+	//
+	// When defined, EndPos must have the same Filename value as the StartPos.
+	EndPos token.Position
+
+	Message string
+	Value   string
+}
+
+// Error implements error.
+func (d Diagnostic) Error() string {
+	return fmt.Sprintf("%s: %s", d.StartPos, d.Message)
+}
+
+// Diagnostics is a collection of diagnostic messages.
+type Diagnostics []Diagnostic
+
+// Add adds an individual Diagnostic to the diagnostics list.
+func (ds *Diagnostics) Add(d Diagnostic) {
+	*ds = append(*ds, d)
+}
+
+// Error implements error.
+func (ds Diagnostics) Error() string {
+	switch len(ds) {
+	case 0:
+		return "no errors"
+	case 1:
+		return ds[0].Error()
+	default:
+		return fmt.Sprintf("%s (and %d more diagnostics)", ds[0], len(ds)-1)
```

Comment:
I'm not sure if this would ever reach the user, but I'd rather we provide a more complete image, as these errors might be related to each other. Should we print out more of them?

Reply:
Do you have a suggestion of what it should return?

Reply:
Going to merge for now, but let's revisit this later and iterate on what the raw error string does

Reply:
Yeah, didn't mean for this to be a blocker of course.

I'm not sure if/where this is going to be presented to the user I was just asking if we should for example, show the 5 top errors (skipping warnings maybe?) before offering the "and N more diagnostics" message.